src/main/java/com/yelp/nrtsearch/server/luceneserver/search/TerminateAfterWrapper.java

```diff
@@ -130,14 +134,30 @@ public TerminateAfterLeafCollectorWrapper(LeafCollector leafCollector) {
 
       @Override
       public void setScorer(Scorable scorer) throws IOException {
-        leafCollector.setScorer(scorer);
+        if (terminatedEarly) {
+          // setup constant scorer to avoid spending time on scoring
+          float score = 1;
+          Weight dummyWeight = new ConstantScoreWeight(new MatchAllDocsQuery(), score) {
+            @Override
+            public boolean isCacheable(LeafReaderContext leafReaderContext) {
+              return true;
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+              return new ConstantScoreScorer(this, this.score(), ScoreMode.COMPLETE_NO_SCORES, DocIdSetIterator.all(context.reader().maxDoc()));
+            }
+          };
+          leafCollector.setScorer(dummyWeight.scorer(currentContext));
+        } else {
+          leafCollector.setScorer(scorer);
```

Comment:
I want you to create new scorable class which has  scorer inside,
and when not terminated early, call scorer.score()
when termianted early, return constant score.

I'm expecting you to define a nested class like below in the TerminateAfterWrapper.java.

```
class TerminatedEarlyScorer extends Scorable {
        private final Scorable in;
        ...
        
        @Override
        public float score() throws IOException{
          if (terminatedEarly) {
            return in.score();
          } else {
            // return the min constant score
          }
        }
      }
```

And use this wrapper to wrap the real scorer and set it.

Reply:
Oh Sorry @taoyyu I thought my implementation is an alternative solution to what we want to do, let me implement your approach and let's see how things go, I'll ping you when this is done

Reply:
The setScorer is only called once or twice with in TerminatedAfterWrapper. You are not able to set it to a different scorer later by checking terminatedEarly. You'll have to create the scorer which can check termiantedEarly internally. You can try run the code with your approach to confirm it.

This is just a demo. You don't need to define the TermiantedEarlyScorer. In Java, you are able to directly new abstract class with override. You can check some java materials for this.

Reply:
thanks for the comments, updated